### PR TITLE
[Discuss] extend tensorflow backend's switch to match theano backend

### PR DESCRIPTION
**Pull request by DingKe**

### CODE REVIEW

**Comment by farizrahman4u** [comment]
It already exists. It's called `K.switch()`.

**Comment by DingKe** [comment]
@farizrahman4u  correct me if I'm wrong, by reading the docstring, K.swtich need a condition if a scalar, not applies element-wise selection.

**Comment by farizrahman4u** [comment]
Just look at the code for K.switch. It does the same thing as you did in this PR. It simply maps to T.switch and tf.select.

**Comment by DingKe** [comment]
For the theano backend, yes. 
But for tensorflow, it's not the case. The tf backend invokes [cond](https://www.tensorflow.org/versions/r0.11/api_docs/python/control_flow_ops.html#cond) which need condition to be scalar.

**Comment by farizrahman4u** [comment]
Then fix K.switch's tensorflow backend.

**Comment by gvtulder** [comment]
My two cents: this seems like a bad idea, because it will break code (such as mine) that depends on `K.switch` being an elementwise operation, which it currently is on Theano.

I agree more with farizrahman4u's suggestion: since `switch` is the name of an existing Theano function and the name doesn't exist in TensorFlow, wouldn't it make more sense to make `K.switch` in TensorFlow similar to the Theano function? You're now inventing a definition of `K.switch` that doesn't match either backend.

You could add a `K.ifelse` or something similar to handle the scalar-condition case. And if you really want optimize `K.switch`, I'd suggest making it check whether the input is a scalar and use `switch` or `ifelse` depending on that (but this seems like an optimization that should be included in Theano).

**Comment by fchollet** [comment]
@gvtulder that's a good point.

**Comment by DingKe** [comment]
@fchollet @gvtulder yes, extending K.switch is much better. But there is still one thing:
theano's switch is basically more powerful than tensorflow's select.
For theano's switch, condition can be a scalar or tensor of the same rank as if/then expression's.
For tensorflow's seleect, condition can be a scalar or vector.


So should we leave it that way (as the new PR).
or should we try to make tf backend does as much as theano backend, which seems nontrivial (th swtich support broadcast of arbitrary dimension).

**Comment by farizrahman4u** [comment]
@DingKe Here is the best thing to do IMO:

* Leave Theano backend for K.switch as is
* Fix TF backend for K.switch so that its behavior matches with Theano backend
* Optional : In theano backend of K.switch, if condition is scalar, use ifelse instead of T.switch

File: keras/backend/theano_backend.py [hunk 1]
```diff
@@ -998,9 +998,22 @@ def _step(input, *states):
...
+        then_expression: TensorFlow operation.
+        else_expression: TensorFlow operation.
     '''
-    return T.switch(condition, then_expression, else_expression)
+    if ndim(condition) == 0:
+        return theano.ifelse.ifelse(condition, then_expression, else_expression)
+    else:
+        dims = ndim(then_expression)
+        while(ndim(condition) < dims):
+            condition = expand_dims(condition)
```
Comment (gvtulder) [inline keras/backend/theano_backend.py]: Is this really useful? It might be more confusing than helpful, because it's making the behaviour of `K.switch` different from that of `theano.tensor.switch`.

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -1328,39 +1328,26 @@ def _cond(condition, then_lambda, else_lambda):
...
+        x_shape = copy.copy(then_expression.get_shape())
+        if condition.dtype != tf.bool:
+            condition = tf.cast(condition, 'bool')
+        x = _cond(condition,
+                  lambda: then_expression,
+                  lambda: else_expression)
+        x.set_shape(x_shape)
+        return x
+    else:
+        return tf.select(condition, then_expression, else_expression)
```
Comment (gvtulder) [inline keras/backend/tensorflow_backend.py]: In your previous version you were casting the condition to `bool`, would you also want to do that here?

**Comment by gvtulder** [comment]
@DingKe: That looks better. I don't really understand why you think Theano's `switch` is much more powerful than `tf.select`, but I'm not a TensorFlow user. The TensorFlow documentation seems to say that `tf.select` also accept tensors the same size as the output, so it should be possible to do an elementwise operation with that.

**Comment by nouiz** [comment]
I don't think it is a good idea to convert by default switch to ifelse for
Theano. Currently, there is some extra overhead as we don't have c code for
it. As we don't have a c interface for "lazy op", it is not trivial to add
it, not super difficult, just not trivial and it isn't a current priority.

**Comment by fchollet** [comment]
TF `select` is getting deprecated and replaced with `where`. So that's what we should be normalizing on.

**Comment by DingKe** [comment]
Basically, tf's switch is aligned to theano's.
When condition is scalar, it's a if/else condition (the old interface); 
Otherwise condition should be the same shape with then/else expression, and a element-wise selection is preformed.
@fchollet 
(@farizrahman4u: not use ifelse for theano according to @nouiz.)

**Comment by gvtulder** [comment]
It looks like the Travis failure is because `tf.where` works differently in TensorFlow v0.11. Perhaps you can make a fallback option that uses `tf.select` if you get this `TypeError`?

**Comment by fchollet** [comment]
Closing outdated PR. If you still care about the content of the PR, please submit a new PR to `master`, updated for the Keras 2.0 API.
